### PR TITLE
adaptable adapters

### PR DESCRIPTION
**Pull request by yongzx**
Screenshot showing training AA on Bloom-1b3
Screen Shot 2022-07-08 at 8 21 52 PM

### CODE REVIEW

**Comment by yongzx** [comment]
```
# axis
LANG="de"
DATA_SAMPLES=1000
VOCAB_SIZE=5000
CH=118500
BIGS_MODEL="bigscience/bloom-1b3"
# BIGS_MODEL="/users/zyong2/data/zyong2/huggingface/bigscience/tr5b-1B3-ckpt118500"

# adapters
ADPT_STRATEGY="aa"
EMBD_SRATEGY="replace"
ADPT_CONFIG="aa"

# tokenizer_dir="/users/zyong2/data/zyong2/bigscience/data/processed/022/tok_${BIGS_MODEL##*/}_${LANG}_oscar_${DATA_SAMPLES}samples_${VOCAB_SIZE}vocab_${EMBD_SRATEGY}"
tokenizer_dir="bigscience/bloom-1b3"
cache_dir="/users/zyong2/data/zyong2/huggingface/"
output_dir="/users/zyong2/data/zyong2/bigscience/data/processed/022/${BIGS_MODEL##*/}_${LANG}_${ADPT_STRATEGY}_${DATA_SAMPLES}samples_${EMBD_SRATEGY}"
logging_dir="/users/zyong2/data/zyong2/bigscience/reports/022/${BIGS_MODEL##*/}_${LANG}_${ADPT_STRATEGY}_${DATA_SAMPLES}samples_${EMBD_SRATEGY}"

mkdir -p $output_dir
mkdir -p $logging_dir

python /users/zyong2/data/zyong2/bigscience/gh/multilingual-modeling/scripts/lang_adapt/madx_run_clm.py \
    --model_name_or_path $BIGS_MODEL \
    --tokenizer_name $tokenizer_dir \
    --dataset_name oscar \
    --cache_dir $cache_dir \
    --dataset_config_name "unshuffled_deduplicated_$LANG" \
    --logging_dir $logging_dir \
    --report_to "tensorboard" \
    --learning_rate 0.001 \
    --do_train \
    --do_eval \
    --output_dir $output_dir \
    --preprocessing_num_workers 8 \
    --overwrite_output_dir \
    --per_device_train_batch_size 2 \
    --gradient_accumulation_steps 4 \
    --per_device_eval_batch_size 2 \
    --eval_accumulation_steps 4 \
    --eval_steps 10 \
    --evaluation_strategy "steps" \
    --max_eval_samples 500 \
    --save_steps 100 \
    --save_strategy "steps" \
    --max_train_samples $DATA_SAMPLES \
    --max_steps 100 \
    --logging_steps 1 \
    --train_adapter \
    --lang_adapt_strategies $ADPT_STRATEGY \
    --embedding_strategies $EMBD_SRATEGY \
    --adapter_config $ADPT_CONFIG \
    --language $LANG \
    --load_best_model_at_end
```

**Comment by haileyschoelkopf** [comment]
Will try running this as soon as I've got GPU space!

Below, kinda for my own reference: 

My interpretation from the paper, though, was that the Gumbel softmax should be choosing between passing the hidden states through 

```
if self.adapter_residual_before_ln:
            output = output + residual_input
            if isinstance(self.config, AAConfig):
                # gumbel switch
                switch = torch.nn.functional.gumbel_softmax(self.pis, tau=self.config.tau, hard=True)
                output = switch[0] * output + switch[1] * residual_input
            else:    
                output = output + residual_input
```

would be: 
```
def forward(self, x, residual_input):
        identity = x
        ...
        if isinstance(self.config, AAConfig):
             # gumbel switch
             switch = torch.nn.functional.gumbel_softmax(self.pis, tau=self.config.tau, hard=True)
             output = switch[0] * output + switch[1] * identity

        ...
        if self.adapter_residual_before_ln:
            output = output + residual_input
```

So adding the residual happens regardless, but there's an additional `identity` residual saved within this `forward()` call so that it can be used in the Gumbel Softmax switch.

**Comment by yongzx** [comment]
Happy to share that the reason for training failure is that the weights go to infinity when we use Rational activation function. 

@haileyschoelkopf you are right that we should move the gumbel part. Furthermore, to ensure stable training, we need to apply layernorm before the adapters bottleneck `ln_before`.
